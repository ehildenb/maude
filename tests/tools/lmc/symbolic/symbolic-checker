#!/bin/sh

MAUDE_LIB=$srcdir/../../../src/Main
export MAUDE_LIB

../../../../src/Main/maude \
  < $srcdir/symbolic-checker.maude -no-banner -no-advise \
  > symbolic-checker.out 2>&1

diff $srcdir/symbolic-checker.expected symbolic-checker.out > /dev/null 2>&1
