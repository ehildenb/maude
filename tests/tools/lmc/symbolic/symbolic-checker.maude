set show timing off .

load ../../../../contrib/systems/cycle.maude
load ../../../../contrib/systems/token.maude
load ../../../../contrib/tools/lmc/symbolic-checker.maude

--- Testing META-NARROWING-INTERFACE
--- ================================

select META-NARROWING-INTERFACE .

--- Function `metaNarrow2`
--- ----------------------

--- ### System `CYCLE`

reduce metaNarrow2(upModule('CYCLE, true), 'f['a.PreState])
    == (    { 'f['b.PreState],'State, none / rl 'f['a.PreState] => 'f['b.PreState] [narrowing label('a2b)] . }
         || { 'g['c.PreState],'State, none / rl 'f['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] . }
       ) .

reduce metaNarrow2(upModule('CYCLE, true), 'f['b.PreState])
    == { 'f['c.PreState],'State, none / rl 'f['b.PreState] => 'f['c.PreState] [narrowing label('b2c)] . } .

--- ### System `BOUNDED-CYCLE`

reduce metaNarrow2(upModule('BOUNDED-CYCLE, true), '`{_`,_`}['1.Nat*, '1.Nat*])
    == { '`{_`,_`}['0.Nat*,'1.NzNat*],'State, none / rl '`{_`,_`}['_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*],'#3:Nat*] => '`{_`,_`}['#1:Nat*,'#3:Nat*] [narrowing label('rm)] . } .

reduce metaNarrow2(upModule('BOUNDED-CYCLE, true), '`{_`,_`}['N:Nat*, '_:+_['1.Nat*, '1.Nat*, '1.Nat*]])
    == (    { '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],                             'State, 'N:Nat* <- '_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*] / rl '`{_`,_`}['_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*],'#3:Nat*]          => '`{_`,_`}['#1:Nat*,'#3:Nat*] [narrowing label('rm)] . }
         || { '`{_`,_`}['1.NzNat*,'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],                            'State, 'N:Nat* <- '0.Nat*                            / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
         || { '`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],           'State, 'N:Nat* <- '0.Nat*                            / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
         || { '`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],           'State, 'N:Nat* <- '1.NzNat*                          / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
         || { '`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]], 'State, 'N:Nat* <- '0.Nat*                            / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
         || { '`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]], 'State, 'N:Nat* <- '1.NzNat*                          / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
         || { '`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]], 'State, 'N:Nat* <- '_:+_['1.NzNat*,'1.NzNat*]         / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
       ) .

reduce metaNarrow2(upModule('BOUNDED-CYCLE, true), '`{_`,_`}['N:Nat*, 'M:Nat*])
    == (    { '`{_`,_`}['#1:Nat*,'M:Nat*],'State,                                                             'N:Nat* <- '_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*]         / rl '`{_`,_`}['_:+_['1.NzNat*, '#1:Nat*,'#2:Nat*],'#3:Nat*]         => '`{_`,_`}['#1:Nat*,'#3:Nat*]                                                                                     [narrowing label('rm)]  . }
         || { '`{_`,_`}['_:+_['1.NzNat*,'N:Nat*,'#3:Nat*],'_:+_['1.NzNat*,'N:Nat*,'#2:Nat*,'#3:Nat*]],'State, 'M:Nat* <- '_:+_['1.NzNat*,'N:Nat*,'#2:Nat*,'#3:Nat*] / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_[ '#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_[ '#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] . }
       ) .

--- Function `allNarrowSteps`
--- -------------------------

--- ### System `EXT-CYCLE`

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'f['a.PreState], 0)
    == allNarrowSteps(upModule('EXT-CYCLE, true), 'f['a.PreState]) .

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'f['a.PreState], 0)
    == (    { 'f['b.PreState],'State,none / rl 'f['a.PreState] => 'f['b.PreState] [narrowing label('a2b)] . }
         || { 'g['c.PreState],'State,none / rl 'f['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
                                            rl 'g['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
            }
       ) .

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'f['a.PreState], 1)
    == ( { 'g['c.PreState],'State,none / rl 'f['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
                                         rl 'g['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
         }
       ) .

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'f['a.PreState], 2)
    == (empty).StepResultSet .

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'f['c.PreState], 0)
    == ( { 'f['a.PreState],'State,none / rl 'f['c.PreState] => 'f['a.PreState] [narrowing label('c2a)] .
                                         rl 'g['c.PreState] => 'g['a.PreState] [narrowing label('c2a)] .
         }
       ) .

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'g['b.PreState], 0)
    == (empty).StepResultSet .

reduce allNarrowSteps(upModule('EXT-CYCLE, true), 'g['a.PreState], 0)
    == ( { 'g['c.PreState],'State,none / rl 'f['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
                                         rl 'g['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
         }
       ) .
