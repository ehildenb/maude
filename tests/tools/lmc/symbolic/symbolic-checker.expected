==========================================
reduce in META-NARROWING-INTERFACE : metaNarrow2(upModule('CYCLE, true), 'f[
    'a.PreState]) == ({'f['b.PreState],'State,none / rl 'f['a.PreState] => 'f[
    'b.PreState] [narrowing label('a2b)] .} || {'g['c.PreState],'State,none /
    rl 'f['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .}) .
rewrites: 1017
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : metaNarrow2(upModule('CYCLE, true), 'f[
    'b.PreState]) == {'f['c.PreState],'State,none / rl 'f['b.PreState] => 'f[
    'c.PreState] [narrowing label('b2c)] .} .
rewrites: 849
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : metaNarrow2(upModule('BOUNDED-CYCLE,
    true), '`{_`,_`}['1.Nat*,'1.Nat*]) == {'`{_`,_`}['0.Nat*,'1.NzNat*],'State,
    none / rl '`{_`,_`}['_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*],'#3:Nat*] =>
    '`{_`,_`}['#1:Nat*,'#3:Nat*] [narrowing label('rm)] .} .
rewrites: 1874
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : metaNarrow2(upModule('BOUNDED-CYCLE,
    true), '`{_`,_`}['N:Nat*,'_:+_['1.Nat*,'1.Nat*,'1.Nat*]]) == (((((({
    '`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,
    '1.NzNat*]],'State,
  'N:Nat* <- '1.NzNat* / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,
    '#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],
    '_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing
    label('add)] .} || {'`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*],'_:+_[
    '1.NzNat*,'1.NzNat*,'1.NzNat*]],'State,
  'N:Nat* <- '_:+_['1.NzNat*,'1.NzNat*] / rl '`{_`,_`}['#1:Nat*,'_:+_[
    '1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_[
    '1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,
    '#2:Nat*]]]] [narrowing label('add)] .}) || {'`{_`,_`}['_:+_['1.NzNat*,
    '1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],'State,
  'N:Nat* <- '0.Nat* / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,
    '#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_[
    '#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label(
    'add)] .}) || {'`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,
    '1.NzNat*,'1.NzNat*]],'State,
  'N:Nat* <- '1.NzNat* / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,
    '#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],
    '_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing
    label('add)] .}) || {'`{_`,_`}['_:+_['1.NzNat*,'1.NzNat*],'_:+_['1.NzNat*,
    '1.NzNat*,'1.NzNat*]],'State,
  'N:Nat* <- '0.Nat* / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,
    '#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_[
    '#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label(
    'add)] .}) || {'`{_`,_`}['1.NzNat*,'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],
    'State,
  'N:Nat* <- '0.Nat* / rl '`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,
    '#3:Nat*]] => '`{_`,_`}['_:+_['#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_[
    '#3:Nat*,'_:+_['1.NzNat*,'_:+_['#1:Nat*,'#2:Nat*]]]] [narrowing label(
    'add)] .}) || {'`{_`,_`}['#1:Nat*,'_:+_['1.NzNat*,'1.NzNat*,'1.NzNat*]],
    'State,
  'N:Nat* <- '_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*] / rl '`{_`,_`}['_:+_['1.NzNat*,
    '#1:Nat*,'#2:Nat*],'#3:Nat*] => '`{_`,_`}['#1:Nat*,'#3:Nat*] [narrowing
    label('rm)] .}) .
rewrites: 5539
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : metaNarrow2(upModule('BOUNDED-CYCLE,
    true), '`{_`,_`}['N:Nat*,'M:Nat*]) == ({'`{_`,_`}['#1:Nat*,'M:Nat*],'State,
  'N:Nat* <- '_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*] / rl '`{_`,_`}['_:+_['1.NzNat*,
    '#1:Nat*,'#2:Nat*],'#3:Nat*] => '`{_`,_`}['#1:Nat*,'#3:Nat*] [narrowing
    label('rm)] .} || {'`{_`,_`}['_:+_['1.NzNat*,'N:Nat*,'#3:Nat*],'_:+_[
    '1.NzNat*,'N:Nat*,'#2:Nat*,'#3:Nat*]],'State,
  'M:Nat* <- '_:+_['1.NzNat*,'N:Nat*,'#2:Nat*,'#3:Nat*] / rl '`{_`,_`}[
    '#1:Nat*,'_:+_['1.NzNat*,'#1:Nat*,'#2:Nat*,'#3:Nat*]] => '`{_`,_`}['_:+_[
    '#1:Nat*,'_:+_['1.NzNat*,'#3:Nat*]],'_:+_['#3:Nat*,'_:+_['1.NzNat*,'_:+_[
    '#1:Nat*,'#2:Nat*]]]] [narrowing label('add)] .}) .
rewrites: 2461
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'f['a.PreState], 0) == allNarrowSteps(upModule('EXT-CYCLE, true), 'f[
    'a.PreState]) .
rewrites: 104
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'f['a.PreState], 0) == ({'f['b.PreState],'State,none / rl 'f['a.PreState]
    => 'f['b.PreState] [narrowing label('a2b)] .} || {'g['c.PreState],'State,
    none / rl 'f['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .
rl 'g['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .}) .
rewrites: 53
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'f['a.PreState], 1) == {'g['c.PreState],'State,none / rl 'f['a.PreState] =>
    'g['c.PreState] [narrowing label('a2c)] .
rl 'g['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .} .
rewrites: 29
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'f['a.PreState], 2) == (empty).StepResultSet .
rewrites: 5
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'f['c.PreState], 0) == {'f['a.PreState],'State,none / rl 'f['c.PreState] =>
    'f['a.PreState] [narrowing label('c2a)] .
rl 'g['c.PreState] => 'g['a.PreState] [narrowing label('c2a)] .} .
rewrites: 29
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'g['b.PreState], 0) == (empty).StepResultSet .
rewrites: 5
result Bool: true
==========================================
reduce in META-NARROWING-INTERFACE : allNarrowSteps(upModule('EXT-CYCLE, true),
    'g['a.PreState], 0) == {'g['c.PreState],'State,none / rl 'f['a.PreState] =>
    'g['c.PreState] [narrowing label('a2c)] .
rl 'g['a.PreState] => 'g['c.PreState] [narrowing label('a2c)] .} .
rewrites: 29
result Bool: true
Bye.
